hazelcast/src/main/java/com/hazelcast/internal/management/ManagementCenterService.java

```diff
@@ -46,36 +53,131 @@
+                // ignore event and clear the queue if the last poll happened a while ago
+                onMCEventWindowExceeded();
+            } else {
+                mcEvents.offer(event);
+            }
+
+        }
+
+        void onMCEventWindowExceeded() {
+            mcEvents.clear();
+            lastAccessTimestamps.clear();
+        }
+
+        /**
+         * @param mcRemoteAddr the address of the calling MC instance.
+         * @return the events which were added to the queue since this MC last polled, or all known events if the MC polls for the
+         * first time.
+         */
+        public List<Event> pollMCEvents(Address mcRemoteAddr) {
+            Long lastAccessObj = lastAccessTimestamps.get(mcRemoteAddr);
+            if (lastAccessObj == null) {
+                ArrayList<Event> copy = new ArrayList<>(mcEvents);
+                updateLatestAccessStats(mcRemoteAddr);
+                return copy;
+            } else {
+                long lastAccess = lastAccessObj;
+                List<Event> recentEvents = new ArrayList<>();
+                for (Event evt : mcEvents) {
+                    if (evt.getTimestamp() >= lastAccess) {
+                        recentEvents.add(evt);
+                    }
+                }
+                updateLatestAccessStats(mcRemoteAddr);
+                return recentEvents;
+            }
+        }
+
+        /**
+         * Updates {@link #lastMCEventsPollMillis} to the current time, removes old entries from {@link #lastAccessTimestamps}
+         * and removes the entries of {@link #mcEvents} that are already read by all known MCs.
+         *
+         * @param mcRemoteAddr
+         */
+        private void updateLatestAccessStats(Address mcRemoteAddr) {
+            lastMCEventsPollMillis = clock.getAsLong();
+            lastAccessTimestamps.put(mcRemoteAddr, lastMCEventsPollMillis);
+            if (mcEvents.isEmpty()) {
+                return;
+            }
+            OptionalLong maybeOldestAccess = lastAccessTimestamps.values().stream().mapToLong(Long::longValue).min();
+            if (maybeOldestAccess.isPresent()) {
+                long oldestAccess = maybeOldestAccess.getAsLong();
+                cleanUpLastAccessTimestamps(oldestAccess);
+                Iterator<Event> it = mcEvents.iterator();
+                while (it.hasNext()) {
+                    Event evt = it.next();
+                    if (evt.getTimestamp() >= oldestAccess) {
+                        break;
+                    }
+                    it.remove();
+                }
+            }
+        }
+
+        /**
+         * Removes the entries from {@link #lastAccessTimestamps} which record accesses older than
+         * {@link #MC_EVENTS_WINDOW_MILLIS}.
+         *
+         * @param oldestAccess
+         */
+        private void cleanUpLastAccessTimestamps(long oldestAccess) {
+            if (lastMCEventsPollMillis - oldestAccess > MC_EVENTS_WINDOW_MILLIS) {
```

Comment:
Right, it is a little optimization to avoid iterating (by `removeIf()`) on the `lastAccessTimestamps` map whenever an MC polls the events. Not very important, but since we don't have prior knowledge about the number of connected MCs, I thought it is better to be cautious. However I accept it is likely to be  premature optimization.